instaclustr/resource_cluster.go

```diff
@@ -601,8 +649,209 @@ func appendIfMissing(slice []string, toAppend string) []string {
+			return "", "", fmt.Errorf("[ERROR] This cluster didn't enable Kibana, kibana_node_sise is not used for this cluster. Please use master_node_size to change the size instead")
+		}
+		changedCount += 1
+		nodePurpose = ELASTICSEARCH_KIBANA
+		nodeSize = kibanaSize
+	}
+	if changedCount > 1 {
+		return "", "", fmt.Errorf("[ERROR] Please change one size at a time or change all to same size at onece")
+	}
+	return nodeSize, nodePurpose, nil
+}
+
+func getBundleOptionKey(bundleIndex int, option string) string {
+	return fmt.Sprintf("bundle.%d.options.%s", bundleIndex, option)
+}
+
+func doElasticsearchClusterResize(client *APIClient, cluster *Cluster, d *schema.ResourceData, bundleIndex int) error {
+	kibana := len(cluster.BundleOption.KibanaNodeSize) > 0
+	dataNodes := len(cluster.BundleOption.DataNodeSize) > 0
+	var nodePurpose *NodePurpose
+	var nodeSize string
+	masterNewSize := getNewSizeOrEmpty(d, getBundleOptionKey(bundleIndex, "master_node_size"))
+	kibanaNewSize := getNewSizeOrEmpty(d, getBundleOptionKey(bundleIndex, "kibana_node_size"))
+	dataNewSize := getNewSizeOrEmpty(d, getBundleOptionKey(bundleIndex, "data_node_size"))
+
+	if newSize, isAllChange := isElasticsearchSizeAllChange(kibanaNewSize, masterNewSize, dataNewSize, kibana, dataNodes); isAllChange {
+		nodeSize = newSize
+		nodePurpose = nil
+	} else {
+		newSize, purpose, err := getSingleChangedElasticsearchSizeAndPurpose(kibanaNewSize, masterNewSize, dataNewSize, kibana, dataNodes)
+		if err != nil {
+			return err
+		}
+		nodeSize = newSize
+		nodePurpose = &purpose
+	}
+	log.Printf("[INFO] Resizing Elasticsearch cluster. nodePurpose: %s, newSize: %s", nodePurpose, nodeSize)
+	err := client.ResizeCluster(cluster.ID, cluster.DataCentres[0].ID, nodeSize, nodePurpose)
+	if err != nil {
+		return fmt.Errorf("[Error] Error resizing cluster %s with error %s", cluster.ID, err)
+	}
+	return nil
+}
 
+func isKafkaSizeAllChange(brokerSize, zookeeperSize string, dedicatedZookeeper bool) (string, bool) {
+	if len(brokerSize) == 0 {
+		return "", false
+	}
+	if dedicatedZookeeper && (len(zookeeperSize) == 0 || zookeeperSize != brokerSize) {
+		return "", false
+	}
+	return brokerSize, true
+}
+
+func getSingleChangedKafkaSizeAndPurpose(brokerSize, zookeeperSize string, dedicatedZookeeper bool) (string, NodePurpose, error) {
+	changedCount := 0
+	var nodePurpose NodePurpose
+	var nodeSize string
+	if len(brokerSize) > 0 {
+		changedCount += 1
+		nodePurpose = KAFKA_BROKER
+		nodeSize = brokerSize
+	}
+	if len(zookeeperSize) > 0 {
+		if !dedicatedZookeeper {
+			return "", "", fmt.Errorf("[ERROR] This cluster didn't enable Dedicated Zookeeper, zookeeper_node_size is not used for this cluster. Please use master_node_size to change the size instead")
+		}
+		nodePurpose = KAFKA_DEDICATED_ZOOKEEPER
+		nodeSize = zookeeperSize
+		changedCount += 1
+	}
+	if changedCount > 1 {
+		return "", "", fmt.Errorf("[ERROR] Please change one size at a time or change all to same size at onece")
```

Comment:
as above:

"[ERROR] Please change either a single node size at a time or change all nodes to the same size at once"

Reply:
Fixed